packages/contracts/contracts/libs/0.8.x/TokenHolderLib.sol

```diff
@@ -0,0 +1,191 @@
+        }
+    }
+
+    /**
+     * @dev Removes holders of specific project tokens.
+     * @param allowedProjectHoldersMapping A mapping of the allowed holders for a specific project.
+     * @param _projectId Project ID to enable minting on.
+     * @param _coreContract The address of the core contract.
+     * @param _registeredNFTAddresses The set of registered NFT addresses.
+     * @param _ownedNFTAddresses NFT core addresses of projects to be removed
+     * from allowlist. Indexes must align with `_ownedNFTProjectIds`.
+     * @param _ownedNFTProjectIds Project IDs on `_ownedNFTAddresses` whose
+     * holders will be removed from allowlist to mint project `_projectId`.
+     * Indexes must align with `_ownedNFTAddresses`.
+     */
+    function removeHoldersOfProjects(
+        mapping(address => mapping(uint256 => mapping(address => mapping(uint256 => bool))))
+            storage allowedProjectHoldersMapping,
+        uint256 _projectId,
+        address _coreContract,
+        EnumerableSet.AddressSet storage _registeredNFTAddresses,
+        address[] memory _ownedNFTAddresses,
+        uint256[] memory _ownedNFTProjectIds
+    ) internal {
+        require(
+            _ownedNFTAddresses.length == _ownedNFTProjectIds.length,
+            "TokenHolderLib: arrays must be same length"
+        );
+        for (uint256 i = 0; i < _ownedNFTAddresses.length; i++) {
+            require(
+                _registeredNFTAddresses.contains(_ownedNFTAddresses[i]),
+                "TokenHolderLib: address not registered"
+            );
+            allowedProjectHoldersMapping[_coreContract][_projectId][
+                _ownedNFTAddresses[i]
+            ][_ownedNFTProjectIds[i]] = false;
+        }
+    }
+
+    /**
+     * @dev Allows and removes holders of specific project tokens in one operation.
+     * @param allowedProjectHoldersMapping A mapping of the allowed holders for a specific project.
+     * @param _projectId The ID of the project to which the holders belong or from which they will be removed.
+     * @param _coreContract The address of the core contract.
+     * @param _registeredNFTAddresses The set of registered NFT addresses.
+     * @param _ownedNFTAddressesAdd NFT core addresses of projects to be
+     * allowlisted. Indexes must align with `_ownedNFTProjectIdsAdd`.
+     * @param _ownedNFTProjectIdsAdd Project IDs on `_ownedNFTAddressesAdd`
+     * whose holders shall be allowlisted to mint project `_projectId`. Indexes
+     * must align with `_ownedNFTAddressesAdd`.
+     * @param _ownedNFTAddressesRemove NFT core addresses of projects to be
+     * removed from allowlist. Indexes must align with
+     * `_ownedNFTProjectIdsRemove`.
+     * @param _ownedNFTProjectIdsRemove Project IDs on
+     * `_ownedNFTAddressesRemove` whose holders will be removed from allowlist
+     * to mint project `_projectId`. Indexes must align with
+     * `_ownedNFTAddressesRemove`.
+     */
+    function allowRemoveHoldersOfProjects(
```

Comment:
Nit: may suggest naming this `updateHoldersOfProjects` or `allowAndRemoveHoldersOfProjects`